runtime/src/configs/mod.rs

```diff
@@ -312,30 +313,55 @@ impl pallet_collator_selection::Config for Runtime {
-const EPOCH_DURATION_IN_BLOCKS: BlockNumber = 10;
-const BLOCKS_BEFORE_RANDOMNESS_VALID: BlockNumber = 3;
-pub struct MockRandomness;
-impl Randomness<H256, BlockNumber> for MockRandomness {
-    fn random(subject: &[u8]) -> (H256, BlockNumber) {
-        // Simple randomness mock that changes each block but its randomness is only valid after 3 blocks
-
-        // Concatenate the subject with the block number to get a unique hash for each block
-        let subject_concat_block = [
-            subject,
-            &frame_system::Pallet::<Runtime>::block_number().to_le_bytes(),
-        ]
-        .concat();
-
-        let hashed_subject = blake2_256(&subject_concat_block);
-
-        (
-            H256::from_slice(&hashed_subject),
-            frame_system::Pallet::<Runtime>::block_number()
-                .saturating_sub(BLOCKS_BEFORE_RANDOMNESS_VALID),
-        )
+/// Only callable after `set_validation_data` is called which forms this proof the same way
+fn relay_chain_state_proof() -> RelayChainStateProof {
+    let relay_storage_root = ParachainSystem::validation_data()
+        .expect("set in `set_validation_data`")
+        .relay_parent_storage_root;
+    let relay_chain_state =
+        ParachainSystem::relay_state_proof().expect("set in `set_validation_data`");
+    RelayChainStateProof::new(ParachainInfo::get(), relay_storage_root, relay_chain_state)
+        .expect("Invalid relay chain state proof, already constructed in `set_validation_data`")
+}
+
+pub struct BabeDataGetter;
+impl pallet_randomness::GetBabeData<u64, Option<Hash>> for BabeDataGetter {
+    // Tolerate panic here because this is only ever called in an inherent (so can be omitted)
+    fn get_epoch_index() -> u64 {
+        if cfg!(feature = "runtime-benchmarks") {
+            // storage reads as per actual reads
+            let _relay_storage_root = ParachainSystem::validation_data();
+            let _relay_chain_state = ParachainSystem::relay_state_proof();
+            const BENCHMARKING_NEW_EPOCH: u64 = 10u64;
+            return BENCHMARKING_NEW_EPOCH;
+        }
+        relay_chain_state_proof()
+            .read_optional_entry(well_known_keys::EPOCH_INDEX)
+            .ok()
+            .flatten()
+            .expect("expected to be able to read epoch index from relay chain state proof")
+    }
+    fn get_epoch_randomness() -> Option<Hash> {
+        if cfg!(feature = "runtime-benchmarks") {
+            // storage reads as per actual reads
+            let _relay_storage_root = ParachainSystem::validation_data();
+            let _relay_chain_state = ParachainSystem::relay_state_proof();
+            let benchmarking_babe_output = Hash::default();
+            return Some(benchmarking_babe_output);
+        }
+        relay_chain_state_proof()
+            .read_optional_entry(well_known_keys::ONE_EPOCH_AGO_RANDOMNESS)
+            .ok()
+            .flatten()
     }
 }
+const EPOCH_DURATION_IN_BLOCKS: BlockNumber = 100;
```

Comment:
Isn't this supposed to be based on the current relay chain (i.e. kusama, polkadot, etc.)?

Reply:
That's the idea, but after researching Cumulus extensively, I couldn't find a way to get the `EPOCH_DURATION_IN_BLOCKS` constant (`babe.epochDuration`) from the relay chain, plus that way it wouldn't be a constant. I think the best we could do would be to extract the constant from the relay chain repo somehow